broid-wechat/src/core/adapter.ts

```diff
@@ -0,0 +1,154 @@
+import WebHookServer from "./webHookServer";
+
+const HTTPOptions: IAdapterHTTPOptions = {
+  host: "127.0.0.1",
+  port: 8080,
+};
+
+export default class Adapter {
+  public serviceID: string;
+
+  private appID: string;
+  private appSecret: string;
+  private client: any;
+  private connected: boolean;
+  private HTTPOptions: IAdapterHTTPOptions;
+  private logLevel: string;
+  private logger: Logger;
+  private parser: Parser;
+  private webhookServer: WebHookServer;
+
+  constructor(obj: IAdapterOptions) {
+    this.serviceID = obj && obj.serviceID || uuid.v4();
+    this.logLevel = obj && obj.logLevel || "info";
+    this.appID = obj && obj.appID;
+    this.appSecret = obj && obj.appSecret;
+
+    this.HTTPOptions = obj.http || HTTPOptions;
+
+    this.logger = new Logger("adapter", this.logLevel);
+
+    if (!this.appID) {
+      throw new Error("appID must be set");
+    }
+    if (!this.appSecret) {
+      throw new Error("appSecret must be set");
+    }
+
+    this.client = Promise.promisifyAll(new WeChat(this.appID, this.appSecret));
+    this.client.setEndpoint("api.wechat.com");
+    this.parser = new Parser(this.client, this.serviceID, this.logLevel);
+  }
+
+  // Return the service ID of the current instance
+  public serviceId(): String {
+    return this.serviceID;
+  }
+
+  public connect(): Observable<Object> {
+    if (this.connected) {
+      return Observable.of({ type: "connected", serviceID: this.serviceId() });
+    }
+
+    this.webhookServer = new WebHookServer(this.serviceID, this.HTTPOptions, this.logLevel);
+    this.webhookServer.listen();
+    this.connected = true;
+
+    return Observable.of(({ type: "connected", serviceID: this.serviceId() }));
+  }
+
+  public disconnect(): Promise<null> {
+    this.connected = false;
+    return this.webhookServer.close();
+  }
+
+  public listen(): Observable<Object> {
+    if (!this.webhookServer) {
+      return Observable.throw(new Error("No webhookServer found."));
+    }
+
+    return Observable.fromEvent(this.webhookServer, "message")
+      .mergeMap((event: Object) => this.parser.parse(event))
+      .mergeMap((parsed: Object | null) => this.parser.validate(parsed))
+      .mergeMap((validated: Object | null) => {
+        if (!validated) { return Observable.empty(); }
+        return Promise.resolve(validated);
+      });
+  }
+
+  public users(): Promise<any | Error> {
+    return this.client.getFollowersAsync()
+      .then((res) => this.client.batchGetUsersAsync(res.data.openid))
+      .then(R.prop("user_info_list"));
+  }
+
+  public send(data: ISendParameters): Promise<Object | Error> {
+    this.logger.debug("sending", { message: data });
+
+    return broidSchemas(data, "send")
+      .then(() => {
+        switch (data.object.type) {
+          case "Note":
+            return this.client.sendTextAsync(data.to.id, data.object.content);
+          case "Audio":
+            return this.uploadFile(data.object.url, "voice", "audio.amr")
```

Comment:
The JSON take optional name field. We can use it at the name of the file with the extension by testing the mime type of the url.